File: 445.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        l1 = self.reverse(l1)
        l2 = self.reverse(l2)
        carry = 0
        l3 = h3 = ListNode(0)
        while l1 or l2 or carry:
            v1 = v2 = 0
            if l1:
                v1 =l1.val
                l1 = l1.next
            if l2:
                v2 = l2.val
                l2 = l2.next
            carry, val = divmod(v1 + v2 + carry, 10)
            l3.next = ListNode(val)
            l3 = l3.next
        return self.reverse(h3.next)
    def reverse(self, head):
        pre = None
        while head:
            pre, pre.next, head = head, pre, head.next
        return pre

File: 445.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        stack<int> q1;
        stack<int> q2;
        while(l1){
            q1.push(l1->val);
            l1 = l1->next;
        }
        while(l2){
            q2.push(l2->val);
            l2 = l2->next;
        }
        int left = 0;
        ListNode* cur = NULL;
        while(!q1.empty() or !q2.empty()){
            int sum = 0;
            if (!q1.empty()){
                sum += q1.top();
                q1.pop();
            }
            if (!q2.empty()){
                sum += q2.top();
                q2.pop();
            }
            int var = (sum + left)%10;
            ListNode* node = new ListNode(var);
            node->next = cur;
            cur = node;
            left = (sum + left) / 10;
        }
        if (left){
            ListNode* node = new ListNode(left);
            node->next = cur;
            cur = node;
        }
        return cur;
    }
};
